app/server/appsmith-plugins/amazons3Plugin/src/main/java/com/external/plugins/AmazonS3Plugin.java

```diff
@@ -164,40 +171,85 @@ public AmazonS3Plugin(PluginWrapper wrapper) {
+                    throw new AppsmithPluginException(
+                            AppsmithPluginError.PLUGIN_EXECUTE_ARGUMENT_ERROR,
+                            "Missing Base64 encoding. When uploading file from a Filepicker widget its Base64 encoded" +
+                            " value must be used - e.g. Filepicker1.files[0].base64. Did you forget to use the Base64" +
+                            " encoded value ?"
+                    );
+                }
+
+                try {
+                    payload = Base64.getDecoder().decode(encodedPayload);
+                } catch (IllegalArgumentException e) {
+                    throw new AppsmithPluginException(
+                            AppsmithPluginError.PLUGIN_ERROR,
+                            "Appsmith has encountered an unexpected error when decoding base64 encoded content. " +
+                            "Please reach out to Appsmith customer support to resolve this."
+                    );
+                }
+            }
+            else {
+                payload = body.getBytes();
+            }
+
+            InputStream inputStream = new ByteArrayInputStream(payload);
             TransferManager transferManager = TransferManagerBuilder.standard().withS3Client(connection).build();
             transferManager.upload(bucketName, path, inputStream, new ObjectMetadata()).waitForUploadResult();
 
-            return true;
+            ArrayList<String> listOfFiles = new ArrayList<>();
+            listOfFiles.add(path);
+            ArrayList<String> listOfUrls = getSignedUrls(connection, bucketName, listOfFiles, durationInMillis);
+            if(listOfUrls.size() != 1) {
+                throw new AppsmithPluginException(
+                        AppsmithPluginError.PLUGIN_ERROR,
+                        "Appsmith has encountered an unexpected error when fetching url from AmazonS3 after file " +
+                        "creation. Please reach out to Appsmith customer support to resolve this."
+                );
+            }
+            String signedUrl = listOfUrls.get(0);
+
+            return signedUrl;
         }
 
         /*
          * - Exception thrown here needs to be handled by the caller.
          */
-        String readFile(AmazonS3 connection, String bucketName, String path) throws IOException {
+        String readFile(AmazonS3 connection, String bucketName, String path, Boolean encodeContent) throws IOException {
             S3Object fullObject = connection.getObject(bucketName, path);
             S3ObjectInputStream content = fullObject.getObjectContent();
+            byte[] bytes = IOUtils.toByteArray(content);
 
-            String result = "";
-            String line = null;
-            BufferedReader reader = new BufferedReader(new InputStreamReader(content));
-            while ((line = reader.readLine()) != null) {
-                result += line + "\n";
+            String result = null;
+            if(encodeContent == true) {
```

Comment:
Same as before. This way of check ensures null values are handled as well.
```suggestion
            if(Boolean.TRUE.equals(encodeContent)) {
```